### PR TITLE
[webkitbugspy] Support automated comments

### PR DESCRIPTION
**Pull request by JonWBedard**
#### 4895716485e1d61d9570bdf9dc39a976201dab5a
<pre>
[webkitbugspy] Support automated comments
<a href="https://bugs.webkit.org/show_bug.cgi?id=235371">https://bugs.webkit.org/show_bug.cgi?id=235371</a>
&lt;rdar://problem/87783426 &gt;

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.user): Add potential credentials to requests.
(Tracker.credentials): Explicitly call credentials from webkitcorepy.
(Tracker._login_arguments): Create login query from credentials.
(Tracker.me): Return user based on credentials.
(Tracker.populate): Add potential credentials to requests.
(Tracker.add_comment): Add comment to bugzilla issue.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.me): Return user based on credentials.
(Tracker.add_comment): Add comment to GitHub issue.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.add_comment): Base class.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/base.py:
(Base.add): We don&apos;t want to edit the original list of comments.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla.__init__): Add optional Environment mock.
(Bugzilla.__enter__): Enter Environment mock.
(Bugzilla.__exit__): Exit Environment mock.
(Bugzilla._user): Use jsonlib instead of json.
(Bugzilla._user_for_credentials): Convert a credential string into
a User object.
(Bugzilla._issue): Use jsonlib instead of json.
(Bugzilla._see_also): Ditto.
(Bugzilla._comments): Ditto.
(Bugzilla._post_comment): Update mock issue database with posted comment.
(Bugzilla.request): Add credentials to each request, handle comment POST request.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/github.py:
(GitHub.__init__): Add optional Environment mock.
(GitHub.__enter__): Enter Environment mock.
(GitHub.__exit__): Exit Environment mock.
(GitHub._post_comment): Update mock issue database with posted comment.
(GitHub.request): Handle comment POST request.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(AppleDirectoryQuery.user_entry_for_attribute_value): Map AppleConnect username
to user.
(RadarModel.CollectionProperty.__init__): Pass parent model in.
(RadarModel.CollectionProperty.add): Add an element to the collection.
(RadarModel.__init__): Pass self into CollectionProperty.
(RadarModel.commit_changes): Update mock issue database with issue contents.
(RadarClient.__init__): Accept authentication_strategy.
(Radar.AuthenticationStrategySystemAccount.__init__): Save username.
(Radar.AuthenticationStrategySystemAccount.username): Return username.
(Radar.AuthenticationStrategySPNego.username): Assume username from environment.
(Radar.DiagnosisEntry): Moved from RadarModel.DiagnosisEntry.
(Radar.transform_user): Add AppleConnect email.
(Radar.__init__): Pass authentication_strategy to RadarClient.
(RadarModel.DiagnosisEntry): Moved to Radar.DiagnosisEntry.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.user): Support queries by AppleConnect username.
(Tracker.me): Use authentication to determine the logged in user.
(Tracker.add_comment): Add comment to Radar.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
(TestBugzilla.test_reference_parse): Use add_comment.
(TestBugzilla.test_me): Added.
(TestBugzilla.test_add_comment): Added.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
(TestGitHub.test_watcher_parse): Use add_comment.
(TestGitHub.test_reference_parse): Ditto.
(TestGitHub.test_me): Added.
(TestGitHub.test_add_comment): Added.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
(TestRadar): Renamed from TestGitHub.
(TestRadar.test_reference_parse): Use add_comment.
(TestRadar.test_me): Added.
(TestRadar.test_add_comment): Added.
(TestGitHub): Renamed TestRadar.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.me): Added.
(Tracker.add_comment): Added.
* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/mocks/__init__.py: Export mock environment, bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/mocks/environment.py: Added.
(Environment): Mock os.environ and reset credential cache.

Canonical link: <a href="https://commits.webkit.org/246228@main">https://commits.webkit.org/246228@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@288312">https://svn.webkit.org/repository/webkit/trunk@288312</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/base.py [hunk 1]
```diff
@@ -56,4 +56,6 @@ def add(self, bug_data):
             if user:
                 self.users.add(type(self).transform_user(user))
 
-        self.issues[id] = {key: value for key, value in bug_data.items()}
+        self.issues[id] = {}
+        for key, value in bug_data.items():
+            self.issues[id][key] = [e for e in value] if isinstance(value, (list, tuple)) else value
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/base.py]: Should we use deep copy instead or we believe we will only have two level nested in `bug_data`?

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py [hunk 2]
```diff
@@ -20,17 +20,19 @@
 # OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-import json
+import json as jsonlib
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py]: I don't see a naming conflict in this file, why do we need this?

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py [hunk 3]
```diff
@@ -84,20 +91,35 @@ def from_string(self, string):
...
+        username, password = self.credentials(required=required)
+        if not username or not password:
+            return '?{}'.format(query) if query else ''
+        return '?login={username}&password={password}{query}'.format(
+            username=username,
+            password=password,
+            query='&{}'.format(query) if query else '',
+        )
+
+    @decorators.Memoize()
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py]: Nit: to keep consistency with `github.py`, maybe use `@webkitcorepy.decorators.Memoize()` instead?

**Review by dewei-zhu — APPROVED** [review]
r=me

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py (hunk 2 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py]: Needed it with a previous organization I used locally, not any more.

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/base.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/base.py]: Forgot about `copy.deepcopy`! That's definitely the right approach.

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/246228@main (4895716485e1d61d9570bdf9dc39a976201dab5a)!

**Pull request merged**